// @prisma/language-server/dist/src/test/linting.test.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const MessageHandler_1 = require("../MessageHandler");
const vscode_languageserver_1 = require("vscode-languageserver");
const assert = __importStar(require("assert"));
const helper_1 = require("./helper");
function assertLinting(expected, fixturePath) {
    return __awaiter(this, void 0, void 0, function* () {
        const document = (0, helper_1.getTextDocument)(fixturePath);
        const diagnosticsResults = yield (0, MessageHandler_1.handleDiagnosticsRequest)(document);
        assert.ok(diagnosticsResults.length != 0);
        expected.forEach((expectedDiagnostic, i) => {
            const actualDiagnostic = diagnosticsResults[i];
            assert.strictEqual(actualDiagnostic.message, expectedDiagnostic.message);
            assert.deepStrictEqual(actualDiagnostic.range, expectedDiagnostic.range);
            assert.strictEqual(actualDiagnostic.severity, expectedDiagnostic.severity);
        });
    });
}
suite('Linting', () => {
    suiteSetup(() => __awaiter(void 0, void 0, void 0, function* () { }));
    const fixturePathMissingArgument = './linting/missingArgument.prisma';
    const fixturePathWrongType = './linting/wrongType.prisma';
    test('Missing argument', () => __awaiter(void 0, void 0, void 0, function* () {
        yield assertLinting([
            {
                message: 'Argument "provider" is missing in data source block "db".',
                range: {
                    start: { line: 0, character: 0 },
                    end: { line: 2, character: 1 },
                },
                severity: vscode_languageserver_1.DiagnosticSeverity.Error,
            },
        ], fixturePathMissingArgument);
    }));
    test('Wrong type', () => __awaiter(void 0, void 0, void 0, function* () {
        yield assertLinting([
            {
                message: 'Type "Use" is neither a built-in type, nor refers to another model, custom type, or enum.',
                range: {
                    start: { line: 14, character: 12 },
                    end: { line: 14, character: 15 },
                },
                severity: vscode_languageserver_1.DiagnosticSeverity.Error,
            },
        ], fixturePathWrongType);
    }));
});
